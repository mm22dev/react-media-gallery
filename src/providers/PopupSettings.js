import React, { Component, createContext } from 'react'
import PropTypes from 'prop-types'
import initData from '../data/popupSettings.json'
import { PRTError } from '../lib/helpers.js'

export const Context = createContext()

export class PopupSettingsProvider extends Component {
  constructor (props) {
    super(props)
    this.state = {
      popupSettings: {
        ...initData.popupSettings,
        ...this.props.PRTSettings
      },
      errors: []
    }
    this.validatePopupSettings = this.validatePopupSettings.bind(this)
    this.handleErrors = this.handleErrors.bind(this)
  }

  validatePopupSettings () {
    const { popupSettings } = this.state
    let errors = []
    const validate = (settingsObj, expectedType) => {
      let errors = []
      for (let key in settingsObj) {
        if (typeof settingsObj[key] !== expectedType && settingsObj[key] !== null) {
          errors.push(`Invalid prop \`${key}\` of type \`${typeof settingsObj[key]}\` supplied to PopReactrox, expected \`${expectedType}\`.`)
        }
      }
      return errors
    }

    // Numbers
    const { baseZIndex, fadeSpeed, overlayOpacity, windowMargin, windowHeightPad, popupSpeed, popupWidth, popupHeight, popupPadding, popupCaptionHeight } = popupSettings
    const numbers = { baseZIndex, fadeSpeed, overlayOpacity, windowMargin, windowHeightPad, popupSpeed, popupWidth, popupHeight, popupPadding, popupCaptionHeight }
    const numErrors = validate(numbers, 'number')
    errors = numErrors.length > 0 ? [...errors, ...numErrors] : [...errors, ...[]]

    // Booleans
    const { popupIsFixed, useBodyOverflow, usePopupEasyClose, usePopupForceClose, usePopupLoader, usePopupCloser, usePopupCaption, usePopupNav, usePopupDefaultStyling, preload } = popupSettings
    const bools = { popupIsFixed, useBodyOverflow, usePopupEasyClose, usePopupForceClose, usePopupLoader, usePopupCloser, usePopupCaption, usePopupNav, usePopupDefaultStyling, preload }
    const boolErrors = validate(bools, 'boolean')
    errors = boolErrors.length > 0 ? [...errors, ...boolErrors] : [...errors, ...[]]

    // Strings
    const { overlayColor, overlayClass, popupBackgroundColor, popupTextColor, popupLoaderTextSize, popupCloserBackgroundColor, popupCloserTextColor, popupCloserTextSize, popupCaptionTextSize, popupBlankCaptionText, popupCloserText, popupLoaderText, popupClass, popupSelector, popupLoaderSelector, popupCloserSelector, popupCaptionSelector, popupNavPreviousSelector, popupNavNextSelector, selector } = popupSettings
    const strings = { overlayColor, overlayClass, popupBackgroundColor, popupTextColor, popupLoaderTextSize, popupCloserBackgroundColor, popupCloserTextColor, popupCloserTextSize, popupCaptionTextSize, popupBlankCaptionText, popupCloserText, popupLoaderText, popupClass, popupSelector, popupLoaderSelector, popupCloserSelector, popupCaptionSelector, popupNavPreviousSelector, popupNavNextSelector, selector }
    const stringErrors = validate(strings, 'string')
    errors = stringErrors.length > 0 ? [...errors, ...stringErrors] : [...errors, ...[]]

    // Functions
    const { onPopupClose, onPopupOpen } = popupSettings
    const functions = { onPopupClose, onPopupOpen }
    const fnErrors = validate(functions, 'function')
    errors = fnErrors.length > 0 ? [...errors, ...fnErrors] : [...errors, ...[]]

    // Caption
    const { caption } = popupSettings
    if (typeof caption !== 'object' && typeof caption !== 'function' && caption !== null) errors.push(`Invalid prop \`caption\` of type \`${typeof caption}\` supplied to PopReactrox, expected \`null\` or \`object\` or \`function\`.`)

    if (errors.length > 0) this.setState({ errors })
  }

  handleErrors () {
    const errors = [...this.state.errors] // create an array copy
    if (errors.length > 0) {
      const currErr = errors[0] // extract first error
      errors.shift() // remove from the head of array
      try {
        throw new PRTError(currErr)
      } catch (err) {
        this.setState({ errors }, console.error(err))
      }
    }
  }

  componentDidMount () {
    this.setState(prevState => ({
      popupSettings: {
        ...prevState.popupSettings,
        popupLoaderSelector: !prevState.popupSettings.usePopupLoader ? null : prevState.popupSettings.popupLoaderSelector,
        popupCloserSelector: !prevState.popupSettings.usePopupCloser ? null : prevState.popupSettings.popupCloserSelector,
        popupCaptionSelector: !prevState.popupSettings.usePopupCaption ? null : prevState.popupSettings.popupCaptionSelector,
        popupNavPreviousSelector: !prevState.popupSettings.usePopupNav ? null : prevState.popupSettings.popupNavPreviousSelector,
        popupNavNextSelector: !prevState.popupSettings.usePopupNav ? null : prevState.popupSettings.popupNavNextSelector
      }
    }), this.validatePopupSettings)
  }

  componentDidUpdate () {
    this.handleErrors()
  }

  render () {
    return (
      <Context.Provider value={{ ...this.state }}>
        {this.props.children}
      </Context.Provider>
    )
  }
}

PopupSettingsProvider.propTypes = {
  PRTSettings: PropTypes.object,
  children: PropTypes.object
}

PopupSettingsProvider.defaultProps = {
  children: null
}

export const connectWithPopupSettings = Container => {
  return class HOC extends Component {
    render () {
      return <Context.Consumer>
        {(context) => <Container {...this.props} {...context} />}
      </Context.Consumer>
    }
  }
}
