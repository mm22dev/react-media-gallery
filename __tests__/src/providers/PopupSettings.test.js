import { PopupSettingsProvider, connectWithPopupSettings } from '../../../src/providers/PopupSettings'
import defaultPopupSettings from '../../../src/data/popupSettings.json'
import { compose } from 'redux'
import toJSON from 'enzyme-to-json'

const { shallow, mount } = Enzyme
const { popupSettings } = defaultPopupSettings

describe('PopupSettings Provider', () => {
  let wrapper,
    _validatePopupSettings

  const spy = jest.spyOn(PopupSettingsProvider.prototype, 'setState')

  it('Rendering Provider component', () => {
    compose(expect, toJSON, shallow)(<PopupSettingsProvider />).toMatchSnapshot()
  })

  describe('Mounting PopupSettings.', () => {
    it('Processes and updates provider values to be served to the consumer.', () => {
      const mockingPopupSettings = {
        usePopupLoader: false,
        usePopupCloser: false,
        usePopupCaption: true,
        usePopupNav: true
      }
      const expectedPopupSettings = {
        ...popupSettings,
        ...mockingPopupSettings,
        popupLoaderSelector: null,
        popupCloserSelector: null
      }
      _validatePopupSettings = jest.fn()
      wrapper = mount(<PopupSettingsProvider PRTSettings={mockingPopupSettings} />)
      wrapper.instance().validatePopupSettings = _validatePopupSettings
      wrapper.instance().componentDidMount()

      expect(wrapper.instance().state.popupSettings).toEqual(expectedPopupSettings)
      expect(_validatePopupSettings).toHaveBeenCalled()
    })
  })

  describe('Validating received props.', () => {
    it('Throws Error when supplied props mismatches type.', () => {
      const mockPRTSettings = {
        baseZIndex: 'mock-error',
        popupIsFixed: 'mock-error',
        overlayColor: 123,
        onPopupClose: 'mock-error',
        caption: 'mock-error'
      }
      wrapper = mount(<PopupSettingsProvider PRTSettings={mockPRTSettings} />)
      wrapper.instance().validatePopupSettings()
      expect(spy.mock.calls[1][0].errors[0]).toBe(`Invalid prop \`baseZIndex\` of type \`string\` supplied to PopReactrox, expected \`number\`.`)
      expect(spy.mock.calls[1][0].errors[1]).toBe(`Invalid prop \`popupIsFixed\` of type \`string\` supplied to PopReactrox, expected \`boolean\`.`)
      expect(spy.mock.calls[1][0].errors[2]).toBe(`Invalid prop \`overlayColor\` of type \`number\` supplied to PopReactrox, expected \`string\`.`)
      expect(spy.mock.calls[1][0].errors[3]).toBe(`Invalid prop \`onPopupClose\` of type \`string\` supplied to PopReactrox, expected \`function\`.`)
      expect(spy.mock.calls[1][0].errors[4]).toBe(`Invalid prop \`caption\` of type \`string\` supplied to PopReactrox, expected \`null\` or \`object\` or \`function\`.`)
    })
  })

  describe('Handling Errors', () => {
    it('Manages errors when component is mounted.', () => {
      const _handleErrors = jest.fn()
      wrapper = mount(<PopupSettingsProvider PRTSettings={{ baseZIndex: 'mock-error' }} />)
      wrapper.instance().handleErrors = _handleErrors
      wrapper.instance().componentDidMount()
      expect(_handleErrors).toHaveBeenCalled()
    })

    it('Deletes errors from state after beeing thrown.', () => {
      let _cdu = jest.fn()
      wrapper = mount(<PopupSettingsProvider />)
      wrapper.instance().componentDidUpdate = _cdu // prevent error deletion
      wrapper.instance().setState({ errors: ['mock-error'] })
      wrapper.instance().handleErrors()
      expect(wrapper.instance().state.errors.length).toBe(0)
    })
  })
})

describe('PopupSettings Consumer', () => {
  let wrapper
  let ComposedComponent
  let MockComponent = () => <div></div>
  ComposedComponent = connectWithPopupSettings(MockComponent)
  wrapper = mount(<PopupSettingsProvider>
    <ComposedComponent />
  </PopupSettingsProvider>)

  it('Serving props to Consumer', () => {
    const expectedPopupSettings = {
      ...popupSettings,
      popupCaptionSelector: null,
      popupNavPreviousSelector: null,
      popupNavNextSelector: null
    }
    expect(wrapper.find(MockComponent).get(0).props.popupSettings).toEqual(expectedPopupSettings) // on Provider's componentDidMount usePopupCaption and usePopupNav (both set to false) force popupCaptionSelector and popupNavSelector to the null value
  })
})
