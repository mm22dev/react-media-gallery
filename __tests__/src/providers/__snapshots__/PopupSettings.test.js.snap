// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopupSettings Provider Rendering Provider component 1`] = `
<ContextProvider
  value={
    Object {
      "errors": Array [],
      "popupSettings": Object {
        "baseZIndex": 1000,
        "caption": null,
        "fadeSpeed": 300,
        "onPopupClose": null,
        "onPopupOpen": null,
        "overlayClass": "popreactrox-overlay",
        "overlayColor": "#000000",
        "overlayOpacity": 0.6,
        "popupBackgroundColor": "#FFFFFF",
        "popupBlankCaptionText": "(untitled)",
        "popupCaptionHeight": 60,
        "popupCaptionSelector": null,
        "popupCaptionTextSize": null,
        "popupClass": "popreactrox-popup",
        "popupCloserBackgroundColor": "#000000",
        "popupCloserSelector": ".closer",
        "popupCloserText": "×",
        "popupCloserTextColor": "#FFFFFF",
        "popupCloserTextSize": "20px",
        "popupHeight": 100,
        "popupIsFixed": false,
        "popupLoaderSelector": ".loader",
        "popupLoaderText": "••••",
        "popupLoaderTextSize": "2em",
        "popupNavNextSelector": null,
        "popupNavPreviousSelector": null,
        "popupPadding": 10,
        "popupSelector": null,
        "popupSpeed": 300,
        "popupTextColor": "#000000",
        "popupWidth": 200,
        "preload": false,
        "selector": null,
        "useBodyOverflow": false,
        "usePopupCaption": false,
        "usePopupCloser": true,
        "usePopupDefaultStyling": true,
        "usePopupEasyClose": true,
        "usePopupForceClose": false,
        "usePopupLoader": true,
        "usePopupNav": false,
        "windowHeightPad": 0,
        "windowMargin": 50,
      },
    }
  }
/>
`;
